File: src/ExpressionAddOperators.java
// Time Complexity : Exponential
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : No
import java.util.ArrayList;
import java.util.List;

public class ExpressionAddOperators {
	// Your code here along with comments explaining your approach
	/*
	 * This approach is by using recursive method but with a for loop.
	 * The + and - operations are performed on the current value chosen and the previously calculated plus and/or minus and respective +/- operations shall be performed
	 * For * a small calculated mess shall be created as previously calculated value - the previous tail value + the previous tail value * the current chosen value
	 */
	//Recursion with for loop
	List<String> result1;
    int targ1;
    public List<String> addOperatorsNormalRecursion(String num, int target) {
    	result1 = new ArrayList<>();
        targ1 = target;
        if(num == null || num.length() == 0) return result1;
        helperNormalRecursion(num, "", 0, 0, 0);
        return result1;
    }
    
    private void helperNormalRecursion(String num, String path, long calc, long tail, int index){
        //base
        if(index == num.length()){
            if(calc == targ){
            	result1.add(path);
            }
            return;
        }
        //logic
        for(int i = index; i < num.length(); i++){
            if(index != i && num.charAt(index) == '0') continue;
            long curr = Long.parseLong(num.substring(index, i+1));
            if(index == 0){
            	helperNormalRecursion(num, path + curr, curr, curr, i + 1);
            }else{
                //+
            	helperNormalRecursion(num, path + "+" + curr, calc + curr, curr, i + 1);
                //-
            	helperNormalRecursion(num, path + "-" + curr, calc - curr, -curr, i + 1);
                //*
            	helperNormalRecursion(num, path + "*" + curr, calc - tail + tail * curr, tail * curr, i + 1);
            }
        }
    }
    
    // Your code here along with comments explaining your approach
 	/*
 	 * This approach is by using backtracking method but with a for loop.
 	 * The + and - operations are performed on the current value chosen and the previously calculated plus and/or minus and respective +/- operations shall be performed
 	 * For * a small calculated mess shall be created as previously calculated value - the previous tail value + the previous tail value * the current chosen value
 	 * finally to back track each time after every operation is performed, the length of the substring taken will be set to the initial value
 	 */
    //Backtracking with for loop
    List<String> result;
    int targ;
    public List<String> addOperatorsBacktracking(String num, int target) {
        result = new ArrayList<>();
        targ = target;
        if(num == null || num.length() == 0) return result;
        helperBacktracking(num, target, new StringBuilder() , 0, 0, 0);
        return result;
    }
    
    private void helperBacktracking(String num, int target, StringBuilder sb, long calc, long tail, int index){
        //base
        if(index == num.length()){
            if(calc == targ){
                result.add(sb.toString());
            }
            return;
        }
        //logic
        for(int i = index; i < num.length(); i++){
            if(index != i && num.charAt(index) == '0') break;
            long curr = Long.parseLong(num.substring(index, i+1));
            int len = sb.toString().length();
            if(index == 0){
            	sb.append(curr);
            	helperBacktracking(num, target, sb, curr, curr, i + 1);
                sb.setLength(len);
            }else{
                //+
            	sb.append("+");
            	sb.append(curr);
            	helperBacktracking(num, target, sb, calc + curr, curr, i + 1);
                sb.setLength(len);
                //-
                sb.append("-");
                sb.append(curr);
                helperBacktracking(num, target, sb,  calc - curr, -curr, i + 1);
                sb.setLength(len);
                //*
                sb.append("*");
                sb.append(curr);
                helperBacktracking(num, target, sb, calc - tail + tail * curr, tail * curr, i + 1);
                sb.setLength(len);
            }
        }
    }
}

File: expression-add-operators/expression-add-operators.cs
public class Solution {
    public IList<string> AddOperators(string num, int target) {
        IList<string> result = new List<string>();
        if(num==null || num.Length==0) return result;
        
        DFS(num,target,new StringBuilder(),result,0,0,0);
        return result;
    }
    
    private void DFS(string num,int target,StringBuilder sb,IList<string> res,int index,long eval,long mul)
    {
        if(index==num.Length)
        {
            if(target==eval)
            {
                res.Add(sb.ToString());
            }
            return;
        }
        
        int len = sb.Length;
        long curr =0;
        
        for(int i=index;i<num.Length;i++)
        {
            if(num[index]=='0' && i!=index)
            {
                break;
            }
            
            curr = 10* curr + (num[i] - '0');
            Console.WriteLine(curr);
            
            if(index==0)
            {
                DFS(num,target,sb.Append(curr),res,i+1,curr,curr);
                sb.Length= len;
            }
            else
            {
                DFS(num,target,sb.Append("+").Append(curr),res,i+1,eval+curr,curr);
                sb.Length=len;
                DFS(num,target,sb.Append("-").Append(curr),res,i+1,eval-curr,-curr);
                 sb.Length=len;
                DFS(num,target,sb.Append("*").Append(curr),res,i+1,eval-mul + mul                                                        *curr,mul*curr);
                sb.Length=len;
​
            }
        }
    }
}
